### PR TITLE
Calculate identifiers for re-exports correctly.

### PR DESCRIPTION
**Pull request by rictic**
- [x] CHANGELOG.md not updated, nothing interesting for a user yet.

### CODE REVIEW

File: src/javascript/javascript-export-scanner.ts [hunk 1]
```diff
@@ -48,24 +48,50 @@ export class Export implements Resolvable, Feature {
...
       astNode: ExportNode, statementAst: babel.Statement,
-      sourceRange: SourceRange|undefined, nodePath: NodePath<babel.Node>) {
+      sourceRange: SourceRange|undefined, nodePath: NodePath<babel.Node>,
+      exportingAllFrom?: Iterable<Export>) {
     this.astNode = astNode;
     this.statementAst = statementAst;
-    for (const name of esutil.getBindingNamesFromDeclaration(astNode)) {
+    let exportedIdentifiers;
+    if (exportingAllFrom !== undefined) {
+      exportedIdentifiers = (function*() {
```
Comment (justinfagnani) [inline src/javascript/javascript-export-scanner.ts]: I don't think we're using IterableX here yet, but `flatMap` would be understandable here for functional folk.

File: src/javascript/javascript-export-scanner.ts [hunk 2]
```diff
@@ -48,24 +48,50 @@ export class Export implements Resolvable, Feature {
...
-    //   resolution algorithm here. Especially in cases of re-export.
-    //   would need to separate out ScannedExport from Export in that case.
+  // TODO: Could potentially get a speed boost by doing the Reference
+  //   resolution algorithm here, rather than re-doing it every single place
+  //   this export is referenced.
+  resolve(document: Document): Feature|undefined {
+    if (babel.isExportAllDeclaration(this.astNode)) {
+      const [import_] =
+          document.getFeatures({kind: 'import', statement: this.statementAst});
+      if (import_ === undefined) {
```
Comment (justinfagnani) [inline src/javascript/javascript-export-scanner.ts]: I'd like to make this not be the case... I think Imports should be features whether they resolve or not, but generate a warning and have an undefined URL. Worth taking not of where we make this assumption.

File: src/test/javascript/javascript-export-scanner_test.ts [hunk 3]
```diff
@@ -40,7 +40,31 @@ suite('JavaScriptExportScanner', () => {
...
+      ['anotherValue'],
+      ['someValue']
     ]);
   });
+
+  test('re-exports across multiple files correctly', async () => {
+    const features = await getExports('javascript/re-export-all.js');
+    // Like the list above, but flattened as they're all exported from one
+    // export statement.
+    assert.containSubset([...features].map((f) => [...f.identifiers]), [[
```
Comment (justinfagnani) [inline src/test/javascript/javascript-export-scanner_test.ts]: is `containSubset` the right check? Why not filter for js-exports and assert set equality?

File: src/javascript/javascript-export-scanner.ts (hunk 1 above)
Comment (rictic) [reply src/javascript/javascript-export-scanner.ts]: Done, added a local definition of flatMap and called that here.

File: src/javascript/javascript-export-scanner.ts (hunk 2 above)
Comment (rictic) [reply src/javascript/javascript-export-scanner.ts]: I'm cool with that. The type checker will save us here when we make that change.

File: src/test/javascript/javascript-export-scanner_test.ts (hunk 3 above)
Comment (rictic) [reply src/test/javascript/javascript-export-scanner_test.ts]: Good catch, this should just be deepEqual. Changed to deepEqual here and elsewhere.

**Pull request merged**